src/SourceGen/src/Util/SyntaxNodeExtensions.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Maui.SourceGen
+{
+	public static class SyntaxNodeExtensions
+	{
+		public static bool TryGetParentSyntax<T>(this SyntaxNode? syntaxNode, out T? result)
+			where T : SyntaxNode
+		{
+			// set defaults
+			result = null;
+
+			if (syntaxNode == null)
+				return false;
+
+			try
+			{
+				syntaxNode = syntaxNode.Parent;
+
+				if (syntaxNode == null)
+					return false;
+
+				if (syntaxNode.GetType() == typeof(T))
+				{
+					result = syntaxNode as T;
+					return true;
+				}
+
+				return TryGetParentSyntax<T>(syntaxNode, out result);
+			}
+			catch
+			{
+				return false;
+			}
+		}
+	}
+
+	public static class TypeDeclarationSyntaxExtensions
+	{
+		const char NESTED_CLASS_DELIMITER = '+';
+		const char NAMESPACE_CLASS_DELIMITER = '.';
+		const char TYPEPARAMETER_CLASS_DELIMITER = '`';
+
```

Comment:
all the type declaration should be prefixed with `global::`, to avoid name clashes with user namespace

Reply:
but you're doing that later on